plugins/telink-dfu/fu-telink-dfu-hid-device.c

```diff
@@ -0,0 +1,370 @@
+#include <linux/hidraw.h>
+#include <linux/input.h>
+#endif
+
+#include "fu-telink-dfu-archive.h"
+#include "fu-telink-dfu-hid-device.h"
+#include "fu-telink-dfu-struct.h"
+
+struct _FuTelinkDfuHidDevice {
+	FuHidDevice parent_instance;;
+};
+
+G_DEFINE_TYPE(FuTelinkDfuHidDevice, fu_telink_dfu_hid_device, FU_TYPE_HID_DEVICE)
+
+#define FU_TELINK_DFU_HID_DEVICE_SHOW_EDP_INFO	0
+#define FU_TELINK_DFU_HID_DEVICE_SHOW_DESC_INFO	0
+#define FU_TELINK_DFU_HID_DEVICE_START_ADDR		0x0000
+#define FU_TELINK_DFU_HID_DEVICE_REPORT_ID		6
+#define FU_TELINK_DFU_HID_DEVICE_OP_CODE		2
+#define FU_TELINK_DFU_HID_DEVICE_REPORT_TIMEOUT	500 /* ms */
+#define FU_TELINK_DFU_HID_DEVICE_OTA_LENGTH		64
+#define FU_TELINK_DFU_HID_DEVICE_PREAMBLE_LEN	6
+#define HID_IFACE	2
+#define HID_EP_IN	(0x80 | 4)
+#define HID_EP_OUT	(0x00 | 5)
+
+static void
+fu_telink_dfu_hid_device_show_edp_info(FuTelinkDfuHidDevice *self, GError **error)
+{
+#if defined(HAVE_HIDRAW_H) && FU_TELINK_DFU_HID_DEVICE_SHOW_EDP_INFO
+	FuUsbDevice *usb_dev = FU_USB_DEVICE(FU_DEVICE(self));
+	FuUsbInterface *iface = NULL;
+	g_autoptr(GPtrArray) ifaces = NULL;
+	g_autoptr(GPtrArray) endpoints = NULL;
+	FuUsbEndpoint *ep;
+	guint ep_dir;
+	guint8 ep_num;
+	guint16 maxpktsize;
+
+	ifaces = fu_usb_device_get_interfaces(usb_dev, error);
+
+	for (guint i = 0; i < ifaces->len; i++) {
+		iface = g_ptr_array_index(ifaces, i);
+		endpoints = fu_usb_interface_get_endpoints(iface);
+		g_debug("intf %u class=%u", i, fu_usb_interface_get_class(iface));
+		for (guint j = 0; j < endpoints->len; j++) {
+			ep = g_ptr_array_index(endpoints, j);
+			ep_dir = fu_usb_endpoint_get_direction(ep);
+			ep_num = fu_usb_endpoint_get_address(ep);
+			maxpktsize = fu_usb_endpoint_get_maximum_packet_size(ep);
+			g_debug("ep=0x%02x(%s), maxpktsize=%u",
+				ep_num,
+				ep_dir == FU_USB_DIRECTION_DEVICE_TO_HOST ? "in" : "out",
+				maxpktsize);
+		}
+	}
+#endif
+}
+
+static void
+fu_telink_dfu_hid_device_show_descriptors(FuTelinkDfuHidDevice *self, GError **error)
+{
+#if defined(HAVE_HIDRAW_H) && FU_TELINK_DFU_HID_DEVICE_SHOW_DESC_INFO
+	FuHidDevice *hid_dev = FU_HID_DEVICE(self);
+	g_autoptr(GPtrArray) descriptors = NULL;
+	g_autoptr(GString) desc_err = g_string_new(NULL);
+
+	descriptors = fu_hid_device_parse_descriptors(hid_dev, error);
+	if (descriptors == NULL)
+		return;
+	g_debug("descriptors=%u", descriptors->len);
+	for (guint i = 0; i < descriptors->len; i++) {
+		g_autoptr(GError) err = NULL;
+		if (desc_err->len > 0) {
+			g_string_append(desc_err, ", ");
+		}
+		g_string_append_printf(desc_err,
+				       "descriptor 0x%x: %s",
+				       i,
+				       err->message);
+	}
+#endif
+}
+
+static gboolean
+fu_telink_dfu_hid_device_setup(FuDevice *device, GError **error)
+{
+	FuTelinkDfuHidDevice *self = FU_TELINK_DFU_HID_DEVICE(device);
+
+	/* HidDevice->setup */
+	if (!FU_DEVICE_CLASS(fu_telink_dfu_hid_device_parent_class)->setup(device, error))
+		return FALSE;
+
+	fu_device_set_name(device, "telink-hid-device");
```

Comment:
name is supposed to be what the user would know -- e.g. the product name like `HD234` -- and it also needs to be in _init() if it's static.

Reply:
resolved in commit 609d10dc01ee587d7d1e64c0485e49a21f368013